x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -185,30 +172,71 @@ func (rw *ReadWriter) WriteTraceEvent(traceID string, id string, event *modelpb.
 
 func (rw *ReadWriter) writeEntry(e *badger.Entry, opts WriterOpts) error {
 	rw.pendingWrites++
-	err := rw.txn.SetEntry(e.WithTTL(opts.TTL))
-	// Attempt to flush if there are 200 or more uncommitted writes.
-	// This ensures calls to ReadTraceEvents are not slowed down;
-	// ReadTraceEvents uses an iterator, which must sort all keys
-	// of uncommitted writes.
-	// The 200 value yielded a good balance between read and write speed:
-	// https://github.com/elastic/apm-server/pull/8407#issuecomment-1162994643
-	if rw.pendingWrites >= 200 {
-		if err := rw.Flush(opts.StorageLimitInBytes); err != nil {
+	entrySize := estimateSize(e)
+	lsm, vlog := rw.s.db.Size()
+
+	// there are multiple ReadWriters writing to the same storage so add
+	// the entry size and consider the new value to avoid TOCTOU issues.
+	pendingSize := rw.s.pendingSize.Add(entrySize)
+	rw.pendingSize += entrySize
+
+	if current := pendingSize + lsm + vlog; opts.StorageLimitInBytes != 0 && current >= opts.StorageLimitInBytes {
+		// flush what we currently have
+		if err := rw.Flush(); err != nil {
+			return err
+		}
+		// Discard the txn and re-create it if the soft limit has been reached.
+		rw.txn.Discard()
+		rw.txn = rw.s.db.NewTransaction(true)
+		return fmt.Errorf("%w (current: %d, limit: %d)", ErrLimitReached, current, opts.StorageLimitInBytes)
+	} else if rw.pendingWrites >= 200 {
+		// Attempt to flush if there are 200 or more uncommitted writes.
+		// This ensures calls to ReadTraceEvents are not slowed down;
+		// ReadTraceEvents uses an iterator, which must sort all keys
+		// of uncommitted writes.
+		// The 200 value yielded a good balance between read and write speed:
+		// https://github.com/elastic/apm-server/pull/8407#issuecomment-1162994643
+		if err := rw.Flush(); err != nil {
 			return err
 		}
 	}
+	err := rw.txn.SetEntry(e.WithTTL(opts.TTL))
+	// the current ReadWriter flushed the transaction and reset the pendingSize so add
+	// the entrySize again.
+	if rw.pendingSize == 0 {
```

Comment:
rw.pendingSize will not be 0. Even after flush it will be baseTransactionSize. Put this block inside rw.pendingWrites >= 200 because it will only be flushed without aborting inside